src/main.rs

```diff
@@ -145,159 +145,117 @@ async fn main() {
+            _ = term_signal.recv() => break,
 
-        tokio::task::spawn(async move {
-            info!("Config autoreloader started");
-
-            loop {
-                interval.tick().await;
-                match reload_config(autoreload_client_server_map.clone()).await {
-                    Ok(changed) => {
-                        if changed {
-                            get_config().show()
-                        }
+            new_client = listener.accept() => {
+                let (socket, addr) = match new_client {
+                    Ok((socket, addr)) => (socket, addr),
+                    Err(err) => {
+                        error!("{:?}", err);
+                        continue;
                     }
-                    Err(_) => (),
                 };
-            }
-        });
-    }
 
-    let mut term_signal = unix_signal(SignalKind::terminate()).unwrap();
-    let mut interrupt_signal = unix_signal(SignalKind::interrupt()).unwrap();
-
-    tokio::select! {
-        // Initiate graceful shutdown sequence on sig int
-        _ = interrupt_signal.recv() => {
-            info!("Got SIGINT, waiting for client connection drain now");
+                let shutdown_rx = shutdown_tx.subscribe();
+                let drain_tx = drain_tx.clone();
+                let client_server_map = client_server_map.clone();
+
+                tokio::task::spawn(async move {
+                    let start = chrono::offset::Utc::now().naive_utc();
+
+                    match client::client_entrypoint(
+                        socket,
+                        client_server_map,
+                        shutdown_rx,
+                        drain_tx,
+                    )
+                    .await
+                    {
+                        Ok(_) => {
+                            let duration = chrono::offset::Utc::now().naive_utc() - start;
+
+                            info!(
+                                "Client {:?} disconnected, session duration: {}",
+                                addr,
+                                format_duration(&duration)
+                            );
+                        }
 
-            // Broadcast that client tasks need to finish
-            shutdown_event_tx.send(()).unwrap();
-            // Closes transmitter
-            drop(shutdown_event_tx);
+                        Err(err) => {
+                            debug!("Client disconnected with error {:?}", err);
+                        }
+                    };
+                });
+            }
 
-            // This is in a loop because the first event that the receiver receives will be the shutdown event
-            // This is not what we are waiting for instead, we want the receiver to send an error once all senders are closed which is reached after the shutdown event is received
-            loop {
-                match tokio::time::timeout(
+            _ = shutdown_rx.recv() => {
```

Comment:
why is there a different branch for this? if the only thing emitting the event is another branch in the select?
https://github.com/levkk/pgcat/pull/144/files?diff=split&w=0#diff-42cb6807ad74b3e201c5a7ca98b911c5fa08380e942be6e4ac5807f8377f87fcR199